### PR TITLE
Improve error handling and abstract download classes

### PR DESCRIPTION
**Pull request by mcaskill**
## Checklist

- [x] I've read the [Contributing page](https://github.com/junaidbhura/composer-wp-pro-plugins/blob/master/CONTRIBUTING.md).
- [x] Issues: #1, #14, #18, #27, #33, #41
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code has proper inline documentation.

## Issue

When an add-on is unsupported or if a download URL can not be retrieved or resolved, an empty string is returned.

This has the unintended consequence of causing Composer to fail much later during the package installation routine resulting in the following error:

#### Example of the error "Failed to extract junaidbhura/ninja-forms-foobar"

```shell
$ composer require junaidbhura/ninja-forms-foobar
./composer.json has been updated
Running composer update junaidbhura/ninja-forms-foobar
Gathering patches from patch file.
Loading composer repositories with package information
Updating dependencies
Nothing to modify in lock file
Installing dependencies from lock file (including require-dev)
Package operations: 1 install, 0 updates, 0 removals
  - Downloading junaidbhura/ninja-forms-foobar (1.0.0)
Gathering patches from patch file.
Gathering patches for dependencies. This might take a minute.
  - Installing junaidbhura/ninja-forms-foobar (1.0.0): Extracting archive
    Failed to extract junaidbhura/ninja-forms-foobar: (9) '/usr/bin/unzip' -qq '…/vendor/composer/tmp-3e5657bb7bb690e28af71c04f2d6fece' -d '…/vendor/composer/a78af123'

[…/vendor/composer/tmp-3e5657bb7bb690e28af71c04f2d6fece]
  End-of-central-directory signature not found.  Either this file is not
  a zipfile, or it constitutes one disk of a multi-part archive.  In the
  latter case the central directory and zipfile comment will be found on
  the last disk(s) of this archive.
unzip:  cannot find zipfile directory in one of …/vendor/composer/tmp-3e5657bb7bb690e28af71c04f2d6fece or
        …/vendor/composer/tmp-3e5657bb7bb690e28af71c04f2d6fece.zip, and cannot find …/vendor/composer/tmp-3e5657bb7bb690e28af71c04f2d6fece.ZIP, period.

    The archive may contain identical file names with different capitalization (which fails on case insensitive filesystems)
    Unzip with unzip command failed, falling back to ZipArchive class
    Install of junaidbhura/ninja-forms-foobar failed

In ZipDownloader.php line 220:

  '…/vendor/composer/tmp-3e5657bb7bb690e28af71c04f2d6fece' is not a zip archive.
```

To summarize this lengthy error: Composer is attempting to unzip a blank file downloaded from the empty string.

This error has confused many, including myself, as to the real cause. There are numerous issues about it.

## Solution

The solution to this is to explicitly throw an exception during the processing of an add-on and the processing of the download URL to abort the package installation process sooner and provide a clear reason for what went wrong.

To simplify the addition of these error messages, I've abstracted common properties and methods into two base classes for greater consistency between plugin downloaders, especially EDD-distributed plugins.

I've chosen [`UnexpectedValueException`](https://www.php.net/class.UnexpectedValueException) since it seems like a good fit for most scenarios based on similar usage in Composer.

There are currently four kinds of error messages thrown:

1. Package is not supported (Ninja Forms and PublishPress Pro)
2. Download URL is invalid/missing (EDD and Gravity Forms)
3. Download version is invalid/missing (EDD only)
4. Download version and installed version do not match (EDD only)

#### Example of the error "Could not find a matching package" (1)

```shell
$ composer require junaidbhura/ninja-forms-foobar
Info from https://repo.packagist.org: #StandWithUkraine
./composer.json has been updated
Running composer update junaidbhura/ninja-forms-foobar
Gathering patches from patch file.
Loading composer repositories with package information
Updating dependencies
Lock file operations: 1 install, 0 updates, 0 removals
  - Locking junaidbhura/ninja-forms-foobar (1.0.0)
Writing lock file
Installing dependencies from lock file (including require-dev)
Package operations: 1 install, 0 updates, 0 removals

In NinjaForms.php line 299:

  Could not find a matching package for junaidbhura/ninja-forms-foobar. Check the package spelling and that the package is supported
```

#### Example of the error "Expected a valid download URL" (2)

```shell
$ composer require junaidbhura/ninja-forms-foobar
./composer.json has been updated
Running composer update junaidbhura/ninja-forms-foobar
Gathering patches from patch file.
Loading composer repositories with package information
Updating dependencies
Nothing to modify in lock file
Installing dependencies from lock file (including require-dev)
Package operations: 1 install, 0 updates, 0 removals

In AbstractEddPlugin.php line 26:

  Expected a valid download URL for package junaidbhura/ninja-forms-foobar
```

#### Example of the error "Expected a valid download version number" (3)

```shell
$ composer require junaidbhura/ninja-forms-foobar
./composer.json has been updated
Running composer update junaidbhura/ninja-forms-foobar
Gathering patches from patch file.
Loading composer repositories with package information
Updating dependencies
Nothing to modify in lock file
Installing dependencies from lock file (including require-dev)
Package operations: 1 install, 0 updates, 0 removals

In AbstractEddPlugin.php line 33:

  Expected a valid download version number for package junaidbhura/ninja-forms-foobar
```

#### Example of the error "Expected download version to match installed version" (4)

```shell
$ composer require junaidbhura/ninja-forms-foobar
./composer.json has been updated
Running composer update junaidbhura/ninja-forms-foobar
Gathering patches from patch file.
Loading composer repositories with package information
Updating dependencies
Nothing to modify in lock file
Installing dependencies from lock file (including require-dev)
Package operations: 1 install, 0 updates, 0 removals

In AbstractEddPlugin.php line 40:

  Expected download version (3.0.0) to match installed version (2.0.0) of package junaidbhura/ninja-forms-foobar
```

## How has this been tested?

I'm testing this on a client project that uses Advanced Custom Fields Pro, ACF Extended Pro, and WPML.

## Screenshots

Screen capture of the error "Failed to extract junaidbhura/ninja-forms-foobar"

Screen capture of the error "Could not find a matching package"

Screen capture of the error "Expected a valid download URL"

Screen capture of the error "Expected a valid download version number"

Screen capture of the error "Expected download version to match installed version"

### CODE REVIEW

File: plugins/AbstractEddPlugin.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+	/**
+	 * Get the download URL for this plugin.
+	 *
+	 * @param  array<string, mixed> $response The EDD API response.
+	 * @return string
+	 */
+	protected function extractDownloadUrl( array $response ) {
+		if ( empty( $response['download_link'] ) || ! is_string( $response['download_link'] ) ) {
+			throw new UnexpectedValueException(sprintf(
```
Comment (junaidbhura) [inline plugins/AbstractEddPlugin.php]: @mcaskill Can we add spaces in here and everywhere in this file to be consistent with WP Coding standards?

File: plugins/GravityForms.php [hunk 2]
```diff
@@ -47,10 +33,15 @@ public function __construct( $version = '', $slug = 'gravityforms' ) {
 	public function getDownloadUrl() {
 		$http     = new Http();
 		$response = unserialize( $http->post( 'https://gravityapi.com/wp-content/plugins/gravitymanager/api.php?op=get_plugin&slug=' . $this->slug . '&key=' . getenv( 'GRAVITY_FORMS_KEY' ) ) );
-		if ( ! empty( $response['download_url_latest'] ) ) {
-			return str_replace( 'http://', 'https://', $response['download_url_latest'] );
+
+		if ( empty( $response['download_url_latest'] ) || ! is_string( $response['download_url_latest'] ) ) {
+			throw new UnexpectedValueException(sprintf(
```
Comment (junaidbhura) [inline plugins/GravityForms.php]: @mcaskill Lets add spaces within parentheses here

File: plugins/NinjaForms.php [hunk 3]
```diff
@@ -321,7 +296,10 @@ public function getDownloadUrl() {
 				break;
 
 			default:
-				return '';
+				throw new UnexpectedValueException(sprintf(
```
Comment (junaidbhura) [inline plugins/NinjaForms.php]: @mcaskill Lets add spaces within parentheses here

File: plugins/PublishPressPro.php [hunk 4]
```diff
@@ -99,7 +84,10 @@ public function getDownloadUrl() {
 				break;
 
 			default:
-				return '';
+				throw new UnexpectedValueException(sprintf(
```
Comment (junaidbhura) [inline plugins/PublishPressPro.php]: @mcaskill Lets add spaces within parentheses here

**Review by junaidbhura — CHANGES_REQUESTED** [review]

**Comment by mcaskill** [comment]
If you prefer, the `AbstractEddPlugin` could be a trait instead, the concept of the abstract classes could be different if you had something in mind, or each exception could be added to each downloader class.

**Comment by junaidbhura** [comment]
@mcaskill Abstract classes may be easier for potential contributors. Lets merge this! 🚀

**Review by junaidbhura — APPROVED** [review]

**Pull request merged**